File: packages/java/anip-service/src/main/java/dev/anip/service/ServiceConfig.java
package dev.anip.service;

import dev.anip.core.CapabilityDeclaration;

import java.util.List;
import java.util.Optional;
import java.util.function.Function;

/**
 * Configuration for an ANIP service instance.
 */
public class ServiceConfig {

    private String serviceId;
    private List<CapabilityDef> capabilities;
    private String storage;   // "sqlite:///path", ":memory:", "postgres://..."
    private String trust;     // "signed" or "anchored"
    private String keyPath;
    private Function<String, Optional<String>> authenticate;
    private ObservabilityHooks hooks;
    private CheckpointPolicy checkpointPolicy;
    private int retentionIntervalSeconds = 60; // default 60, -1 to disable

    public ServiceConfig() {}

    public String getServiceId() {
        return serviceId;
    }

    public ServiceConfig setServiceId(String serviceId) {
        this.serviceId = serviceId;
        return this;
    }

    public List<CapabilityDef> getCapabilities() {
        return capabilities;
    }

    public ServiceConfig setCapabilities(List<CapabilityDef> capabilities) {
        this.capabilities = capabilities;
        return this;
    }

    public String getStorage() {
        return storage;
    }

    public ServiceConfig setStorage(String storage) {
        this.storage = storage;
        return this;
    }

    public String getTrust() {
        return trust;
    }

    public ServiceConfig setTrust(String trust) {
        this.trust = trust;
        return this;
    }

    public String getKeyPath() {
        return keyPath;
    }

    public ServiceConfig setKeyPath(String keyPath) {
        this.keyPath = keyPath;
        return this;
    }

    public Function<String, Optional<String>> getAuthenticate() {
        return authenticate;
    }

    public ServiceConfig setAuthenticate(Function<String, Optional<String>> authenticate) {
        this.authenticate = authenticate;
        return this;
    }

    public ObservabilityHooks getHooks() {
        return hooks;
    }

    public ServiceConfig setHooks(ObservabilityHooks hooks) {
        this.hooks = hooks;
        return this;
    }

    public CheckpointPolicy getCheckpointPolicy() {
        return checkpointPolicy;
    }

    public ServiceConfig setCheckpointPolicy(CheckpointPolicy checkpointPolicy) {
        this.checkpointPolicy = checkpointPolicy;
        return this;
    }

    public int getRetentionIntervalSeconds() {
        return retentionIntervalSeconds;
    }

    public ServiceConfig setRetentionIntervalSeconds(int retentionIntervalSeconds) {
        this.retentionIntervalSeconds = retentionIntervalSeconds;
        return this;
    }
}

File: packages/csharp/src/Anip.Service/ServiceConfig.cs
using Anip.Core;

namespace Anip.Service;

/// <summary>
/// Configuration for an ANIP service.
/// </summary>
public class ServiceConfig
{
    public string ServiceId { get; set; } = "";
    public List<CapabilityDef> Capabilities { get; set; } = new();

    /// <summary>"sqlite:///path" or ":memory:" (default).</summary>
    public string Storage { get; set; } = ":memory:";

    /// <summary>"signed" or "anchored".</summary>
    public string Trust { get; set; } = "signed";

    /// <summary>Path for key persistence. Null/empty for in-memory keys.</summary>
    public string? KeyPath { get; set; }

    /// <summary>
    /// Bootstrap authentication function. Takes a bearer token string,
    /// returns the principal string or null if not authenticated.
    /// </summary>
    public Func<string, string?>? Authenticate { get; set; }

    /// <summary>Observability hooks (optional).</summary>
    public ObservabilityHooks? Hooks { get; set; }

    /// <summary>Automatic checkpoint scheduling. Null disables.</summary>
    public CheckpointPolicy? CheckpointPolicy { get; set; }

    /// <summary>
    /// How often the retention sweep runs in seconds (default 60).
    /// Set to -1 to disable automatic retention enforcement.
    /// </summary>
    public int RetentionIntervalSeconds { get; set; } = 60;

    /// <summary>
    /// Event class to tier to duration mapping.
    /// Null uses defaults.
    /// </summary>
    public RetentionPolicy? RetentionPolicy { get; set; }

    /// <summary>
    /// Controls failure detail disclosure.
    /// One of "full", "reduced", "redacted", "policy". Default: "full".
    /// </summary>
    public string DisclosureLevel { get; set; } = "full";

    /// <summary>
    /// Maps caller classes to disclosure levels.
    /// Only used when DisclosureLevel is "policy".
    /// </summary>
    public Dictionary<string, string>? DisclosurePolicy { get; set; }

    /// <summary>
    /// Aggregation window size in seconds.
    /// 0 or negative disables aggregation. Default: 0 (disabled).
    /// </summary>
    public int AggregationWindowSeconds { get; set; }
}
